google/cloud/storage/internal/xml_node.cc

```diff
@@ -63,232 +80,235 @@ std::string UnescapeTextContent(absl::string_view text_content) {
+    unparsed_.remove_prefix(1);
   }
-  auto xml_decl_end_pos =
-      content_.find(kXmlDeclEnd, xml_decl_start_pos + kXmlDeclStart.size());
-  if (xml_decl_end_pos == std::string::npos) {
-    return internal::InvalidArgumentError("XML declaration doesn't end",
-                                          GCP_ERROR_INFO());
+
+  auto top = std::move(node_stack_.top());
+  node_stack_.pop();
+  if (!node_stack_.empty()) {
+    return internal::InvalidArgumentError(
+        absl::StrCat("Unterminated tag '", top->GetTagName(), "'"),
+        GCP_ERROR_INFO());
   }
-  i_ = xml_decl_end_pos + kXmlDeclEnd.size();
-  return Status();
+  return top;
 }
 
-Status XmlParser::MainLoop() {
-  auto c = content_[i_];
-  if (c == '<') {
-    auto result = HandleLt();
-    if (!result.ok()) return result;
-  } else if (state_ == ParseState::kStartTag) {
-    auto result = HandleStartTag(c);
-    if (!result.ok()) return result;
-  } else if (state_ == ParseState::kReadingTag) {
-    auto result = HandleReadingTag(c);
-    if (!result.ok()) return result;
-  } else if (state_ == ParseState::kReadingAttr) {
-    auto result = HandleReadingAttr(c);
-    if (!result.ok()) return result;
-  } else if (state_ == ParseState::kEndTag) {
-    HandleEndTag(c);
-  } else if (state_ == ParseState::kReadingText) {
-    // Append the character until we see the next '<'.
-    text_content_ += c;
-  } else if (state_ == ParseState::kBeginClosingTag) {
-    auto result = HandleBeginClosingTag(c);
-    if (!result.ok()) return result;
-  } else if (state_ == ParseState::kReadingClosingTag) {
-    auto result = HandleReadingClosingTag(c);
-    if (!result.ok()) return result;
+// Should skip this during the parse instead.
+Status XmlParser::SkipDeclaration() {
+  constexpr static absl::string_view kDeclStart = "<?xml";
+  constexpr static absl::string_view kDeclEnd = "?>";
+  // Note: This also skips anything before the declaration.
+  auto decl_start = unparsed_.find(kDeclStart);
+  if (decl_start != std::string::npos) {
+    auto decl_end = unparsed_.find(kDeclEnd, decl_start + kDeclStart.size());
+    if (decl_end == std::string::npos) {
+      return internal::InvalidArgumentError("Unterminated XML declaration",
+                                            GCP_ERROR_INFO());
+    }
+    unparsed_.remove_prefix(decl_end + kDeclEnd.size());
   }
   return Status();
 }
 
-Status XmlParser::HandleLt() {
-  if (state_ == ParseState::kReadingText) {
-    // The parser was reading text part. Add a text node to the current
-    // parent if the limits allow.
-    auto text_node =
-        AppendTextNode(UnescapeTextContent(StripTrailingSpaces(text_content_)));
-    if (!text_node) return std::move(text_node).status();
+Status XmlParser::HandleBase() {
+  auto c = unparsed_.front();
+  if (!IsSpace(c)) {
+    if (c != '<') {
+      return internal::InvalidArgumentError(
+          absl::StrCat("Expected tag but found '", unparsed_.substr(0, 4), "'"),
```

Comment:
I am guessing we include the next 4 characters because our test string `"text"` is 4 characters? :/

I am not sure what to return though. Maybe we also include the size of `unparsed_`, which would give an exact position of the error?

Reply:
> I am guessing we include the next 4 characters because our test string `"text"` is 4 characters? :/

No ... I used to have it as only 1, but decided, from the look of the error message, that that didn't give enough context.  4 was just my next best guess.

> Maybe we also include the size of `unparsed_`, which would give an exact position of the error?

I did have `(unparsed_.data() - document_.data())` giving a position for a while, but soon realized this was mostly useless because of the RE stuff we do up front.  That is, the position we would give is probably unrelated to the position in the original string.